`ct/l1`: add `log_collector` and `log_sampler` [CT - Compaction #5]

WillemKauf opened the pull request:
Added here are the `log_collector` and `log_sampler`.

The `log_collector` is a barebones abstract class at this point with only two functions that need to be implemented (`start_collecting_logs()` and `stop_collecting_logs()`).  It is up to the implementer to make this class responsible for pushing CTPs/logs that require compaction to the `compaction_scheduler` (still not present in the tree) for managing, as well as their removal.

The `log_sampler` is responsible for issuing `get_compaction_info()` requests to the `metastore` when attempting to schedule a round of compactions. This class does not make any decisions about whether a `log` needs compacting or not, nor does it filter out sampled logs that may not need compaction in its returned container from `sample_logs()`. Scheduling decisions such as those are left to the other components in the `cloud_topics` compaction subsystem.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

WillemKauf committed:
`ct/l1`: add `log_collector`

Abstract class for implementing a push based `log_collector` for registering
and degestering CTPs/logs that are eligible for compaction on this broker to
the `compaction_scheduler`.

WillemKauf committed:
`ct/l1`: add `log_sampler`

Responsible for grabbing `compaction_info`s from the `metastore`.

WillemKauf committed:
`ct/l1`: add `log_sampler_test`

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--cbebe619-442e-4f7a-9b45-ba0685850b78-->
<!--TEST RESULTS-->
<!--95fb93f6-ef7c-4ea8-ad09-3533c7a3d589-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72602">build#72602</a></summary>


| test_class                           | test_method                                                     | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                                       |
|:-------------------------------------|:----------------------------------------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync                                      | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/72602#01996303-b35d-441e-a934-a505bc70a7ca | FLAKY         | 17/21    | upstream reliability is '91.28787878787878'. current run reliability is '80.95238095238095'. drift is 10.3355 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync                      |
| ClusterRateQuotaTest                 | test_client_response_throttle_mechanism                         | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72602#019962fc-49e1-4b11-acda-4073a93ec8d5 | FLAKY         | 11/21    | upstream reliability is '97.51243781094527'. current run reliability is '52.38095238095239'. drift is 45.13149 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism                         |
| ClusterRateQuotaTest                 | test_client_response_throttle_mechanism_applies_to_next_request | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72602#01996303-b35a-43bb-8f69-22d3c5d358d7 | FLAKY         | 14/21    | upstream reliability is '100.0'. current run reliability is '66.66666666666666'. drift is 33.33333 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
| DataMigrationsApiTest                | test_creating_and_listing_migrations                            | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72602#01996303-b357-49bf-9619-cca951908f07 | FLAKY         | 20/21    | upstream reliability is '99.13494809688581'. current run reliability is '95.23809523809523'. drift is 3.89685 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations                           |
| NodeWiseRecoveryTest                 | test_recovery_local_data_missing                                | {"wait_for_final_manifest_uploads": true}                                    | integration | https://buildkite.com/redpanda/redpanda/builds/72602#019962fc-49e3-4114-8db8-a0631b0c4c06 | FLAKY         | 11/21    | upstream reliability is '97.77227722772277'. current run reliability is '52.38095238095239'. drift is 45.39132 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodeWiseRecoveryTest&test_method=test_recovery_local_data_missing                                |
| SchemaRegistryAutoAuthTest           | test_restarts                                                   | {"move_controller_leader": true}                                             | integration | https://buildkite.com/redpanda/redpanda/builds/72602#019962fc-49df-4f9e-b996-056284d9aa37 | FLAKY         | 20/21    | upstream reliability is '99.74025974025975'. current run reliability is '95.23809523809523'. drift is 4.50216 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SchemaRegistryAutoAuthTest&test_method=test_restarts                                             |
| TopicRecoveryTest                    | test_prevent_recovery                                           | {"cloud_storage_type": 2}                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/72602#01996303-b356-4916-8fcf-a7eb2828c148 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TopicRecoveryTest&test_method=test_prevent_recovery                                              |
| TopicRecoveryTest                    | test_prevent_recovery                                           | {"cloud_storage_type": 1}                                                    | integration | https://buildkite.com/redpanda/redpanda/builds/72602#019962fc-49df-4f9e-b996-056284d9aa37 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TopicRecoveryTest&test_method=test_prevent_recovery                                              |
| WriteCachingFailureInjectionTest     | test_unavoidable_data_loss                                      | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72602#019962fc-49e0-4690-871c-1297c7f9a47b | FLAKY         | 18/21    | upstream reliability is '97.41935483870968'. current run reliability is '85.71428571428571'. drift is 11.70507 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss                          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72772">build#72772</a></summary>


| test_class                           | test_method                                                     | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                                       |
|:-------------------------------------|:----------------------------------------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AvailabilityTests                    | test_availability_when_one_node_failed                          | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72772#01997829-81de-444b-a194-48960dcffd29 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=AvailabilityTests&test_method=test_availability_when_one_node_failed                             |
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync                                      | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/72772#01997829-81df-4d86-bb4b-db7f90c28ea8 | FLAKY         | 20/21    | upstream reliability is '91.37529137529138'. current run reliability is '95.23809523809523'. drift is -3.8628 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync                      |
| ClusterRateQuotaTest                 | test_client_response_and_produce_throttle_mechanism             | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72772#01997829-a7a5-45f3-b018-c50224c3e79a | FLAKY         | 16/21    | upstream reliability is '84.19811320754717'. current run reliability is '76.19047619047619'. drift is 8.00764 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_and_produce_throttle_mechanism             |
| ClusterRateQuotaTest                 | test_client_response_throttle_mechanism                         | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72772#01997829-81dc-4c5a-827c-1e74880a8771 | FLAKY         | 9/21     | upstream reliability is '93.39622641509435'. current run reliability is '42.857142857142854'. drift is 50.53908 and the allowed drift is set to 50. The test should FAIL | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism                         |
| ClusterRateQuotaTest                 | test_client_response_throttle_mechanism_applies_to_next_request | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72772#01997829-a7a7-4b57-8ff1-043b29b51140 | FLAKY         | 12/21    | upstream reliability is '86.19354838709677'. current run reliability is '57.14285714285714'. drift is 29.05069 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
| DisablingPartitionsTest              | test_disable                                                    | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72772#01997829-81e0-4fb5-bfa8-ff24c0251ad2 | FLAKY         | 14/21    | upstream reliability is '86.09112709832134'. current run reliability is '66.66666666666666'. drift is 19.42446 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DisablingPartitionsTest&test_method=test_disable                                                 |
</details>
<!--/body-->

WillemKauf committed:
`config`: make `delete_retention_ms` default `24h`

With a legacy default, since mixed version tests will fail with the new
default due to validation performed in early versions of `redpanda` on
`delete.retention.ms` and tiered storage properties enablement being mutually
exclusive.

WillemKauf commented on src/v/cloud_topics/level_one/compaction/log_sampler.cc:
```diff
@@ -0,0 +1,100 @@
+            continue;
+        }
+
+        auto& sample = samples.at(log.tidp);
```
> Using at() can throw an exception if the key is not found. Since the comment states that not every log in log_list_t is guaranteed to have an entry in samples, this should use find() or contains() to safely check for existence first.

medium bad bot. Comment states that not every log in `logs_list_t` is guaranteed to have a sample in the return value of `log_sampler::sample_logs()`- but it is expected that every log has a sample in `samples` returned from `_metastore->get_compaction_infos(to_sample)`, whether that is a result or an error.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72772-->
<!--fc75eadc-2fb0-4c6a-ad18-0ef685af80d6-->
<!--title-->
## Retry command for [Build#72772](https://buildkite.com/redpanda/redpanda/builds/72772)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_quota_test.py::ClusterRateQuotaTest.test_client_response_throttle_mechanism
```
<!--/body-->

WillemKauf commented:
Closing because this is outdated with changes in my local branch. I will put up another PR in the future with these changes.
